optaplanner-core/src/test/java/org/optaplanner/core/impl/solver/termination/UnimprovedTimeMillisSpentScoreDifferenceThresholdTerminationTest.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.solver.termination;
+
+import java.time.Clock;
+
+import org.junit.Test;
+import org.optaplanner.core.api.score.buildin.simple.SimpleScore;
+import org.optaplanner.core.impl.localsearch.scope.LocalSearchPhaseScope;
+import org.optaplanner.core.impl.localsearch.scope.LocalSearchStepScope;
+import org.optaplanner.core.impl.phase.scope.AbstractPhaseScope;
+import org.optaplanner.core.impl.phase.scope.AbstractStepScope;
+import org.optaplanner.core.impl.solver.scope.DefaultSolverScope;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.assertj.core.api.Assertions.withPrecision;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class UnimprovedTimeMillisSpentScoreDifferenceThresholdTerminationTest {
+
+    private static final long START_TIME_MILLIS = 0L;
+
+    @Test
+    public void forZeroUnimprovedTimeMillis_exceptionIsThrown() {
+        assertThatIllegalArgumentException()
+                .isThrownBy(() -> new UnimprovedTimeMillisSpentTermination(0L))
+                .withMessageContaining("cannot be negative");
```

Comment:
Slightly confusing message. Although I agree it might depend on the context, it seems that in most cases, zero is considered to be neither negative nor positive. Let's use an unambiguous message, for example "must be greater than zero" or "must not be negative or zero".

It's already been [discussed](https://github.com/kiegroup/optaplanner/pull/449#discussion_r227286991) and [dealt with](e1017c77a3a25598ce2491ce1fafd3efae372e40) in the past.

Useful resources:
- http://mathworld.wolfram.com/Zero.html
- https://math.stackexchange.com/questions/26705/is-zero-positive-or-negative

Reply:
That's a good point, the error messages are not clear enough in this regard. Will improve them.

Reply:
Interestingly, the `UnimprovedTimeMillisSpentTermination` checks only for negative numbers (zero is allowed). @ge0ffrey is this difference expected?

There is the same kind of difference, i.e. `>` vs. `>=` also in the `isTerminated()` method.

Reply:
The acceptance of zero was intentional IIRC. It's a general philosophy: 0 time available, empty entity list, etc - it happens in dynamic usage of optaplanner and it's strange but not invalid. In fact, we violate that philisophy a few times (empty customer list for vrp IIRC) and it's a big pita, but we intend to fix those.

Reply:
+1 , I will adjust the fail-fast checks to allow `0` millis.